File: Data Structures & Algorithms/valid-tree/submission-0.py
class Solution:
    def validTree(self, n: int, edges: List[List[int]]) -> bool:
        if len(edges) != n-1:
            return False
        
        self.parent = list(range(n))
        self.rank = [1] *n
        for u,v in edges:
            if not self.union(u,v):
                return False
        return True
    
    def find(self,x):
        if self.parent[x] != x:
            self.parent[x] = self.find(self.parent[x])
        
        return self.parent[x]
    
    def union(self,x,y):
        rootX = self.find(x)
        rootY = self.find(y)

        if rootX == rootY:
            return False
        
        if self.rank[rootX] < self.rank[rootY]:
            rootX,rootY =  rootY,rootX
        
        self.parent[rootY] = rootX

        if self.rank[rootX] == self.rank[rootY]:
            self.rank[rootX]+=1
        
        return True

File: Data Structures & Algorithms/count-connected-components/submission-0.java
class Solution {
     int[] parent;
    public int countComponents(int n, int[][] edges) {
        int count = n;

         parent = new int[n];

         for (int i =0; i<n; i++){
            parent[i] = i;
         }
        for (int[] edge : edges){
            int x = edge[0];
            int y = edge[1];

            if (find(x) != find(y)){
                union(x,y);
                count--;
            }
        }
        return count;


    }

    private int find(int x){
        if (parent[x] !=x){
            parent[x] = find(parent[x]);
        }

        return parent[x];
    }

    private void union(int x, int y){
        int rootX = find(x);
        int rootY = find(y);

        parent[rootX] = rootY;
    }
}
